djangocms_picture/models.py

```diff
@@ -1,83 +1,233 @@
+    def __str__(self):
+        if self.picture and self.picture.label:
+            return self.picture.label
+        return str(self.link_url or self.pk)
 
-    longdesc = models.CharField(
-        _("long description"), max_length=255, blank=True, null=True,
-        help_text=_("When user hovers above picture, this text will appear "
-                    "in a popup."))
+    def get_size(self):
+        # automatic scaling
+        width = 0
+        height = 0
+        crop = False
+        upscale = False
 
-    float = models.CharField(
-        _("side"), max_length=10, blank=True, null=True, choices=FLOAT_CHOICES,
-        help_text=_("Move image left, right or center."))
+        # use width and height attributes
+        if self.width:
+            width = self.width
+        if self.height:
+            height = self.height
+        # use field cropping options
+        if self.use_crop:
+            crop = self.use_crop
+        if self.use_upscale:
+            upscale = self.use_upscale
+        # use field thumbnail settings
+        if self.use_thumbnail:
+            width = self.use_thumbnail.width
+            height = self.use_thumbnail.height
+            crop = self.use_thumbnail.crop
+            upscale = self.use_thumbnail.upscale
 
-    width = models.IntegerField(_("width"), blank=True, null=True,
-                                help_text=_("Pixel"))
-    height = models.IntegerField(_("height"), blank=True, null=True,
-                                 help_text=_("Pixel"))
+        options = {
+            'size': (width, height),
+            'crop': crop,
+            'upscale': upscale,
+        }
+        return options
 
-    def __str__(self):
-        if self.alt:
-            return self.alt[:40]
-        elif self.image:
-            # added if, because it raised attribute error when file wasn't
-            # defined.
-            try:
-                return u"%s" % os.path.basename(self.image.name)
-            except AttributeError:
-                pass
-        return u"<empty>"
+    def get_link(self):
+        if self.link_url:
+            return self.link_url
+        if self.link_page:
+            return self.link_page.get_absolute_url()
+        return False
 
     def clean(self):
-        if self.url and self.page_link:
+        # there can be only one link type
+        if self.link_url and self.link_page:
             raise ValidationError(
-                _("You can enter a Link or a Page, but not both."))
+                ugettext('You have defined an external and internal link. '
+                    'Only one option is allowed.')
+            )
+        # you shall only set one image kind
+        if not self.picture and not self.external_picture:
+            raise ValidationError('You need to add either an image or an external link.')
+        # certain cropping settings don't play nice with one another
+        if (self.use_automatic_scaling and self.use_no_cropping or
```

Comment:
Yes, please rewrite as:

Set this on the model itself:

```python
invalid_option_pairs = [
    ('use_automatic_scaling', 'use_no_cropping'),
    ('use_automatic_scaling', 'use_crop'),
    ('use_automatic_scaling', 'use_upscale'),
    ('use_automatic_scaling', 'use_thumbnail'),
    ('use_no_cropping', 'use_crop'),
    ('use_no_cropping', 'use_upscale'),
    ('use_no_cropping', 'use_thumbnail'),
    ('use_thumbnail', 'use_crop'),
    ('use_thumbnail', 'use_upscale'),
]
```

Then on the clean method do:

```python
invalid_option_pair = None

for pair in invalid_option_pairs:
    if getattr(self, pair[0]) and getattr(self, pair[1]):
        invalid_option_pair = pair
        break


if invalid_option_pair:
    field_1 = self._meta.get_field(invalid_option_pair[0])
    field_2 = self._meta.get_field(invalid_option_pair[1])
    message = ugettext('The cropping selection is not valid. '
                       'You cannot combine %(field_1)s and %(field_2)s.')
    message = message % {'field_1': field_1.verbose_name, 'field_2': field_2.verbose_name}
    raise ValidationError(message)
```

you might have to call force_text() on the verbose names, haven't tested.